design/hardware-status.md

```diff
@@ -0,0 +1,157 @@
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [References](#references)
+
+## Summary
+
+This document explains how information about hardware components (drives, power
+supplies, fans, etc.) including health data, will be retrieved, collected and
+monitored to provide constant up-to-date status on hardware devices present in
+a server.
+The base idea behind this is having a service running in each active host that
+will regularly collect the data and post it to ironic-inspector, from where it
+can be consumed as part of the regular polling for status.
+
+## Motivation
+
+We need to be able to collect and update enough information about hardware
+devices in a server to monitor their status and identify them for
+maintenance and replacement.
+Valuable information are for example the current status, type, serial number,
+physical location, part number.
+We need to be able to refresh the information about the status of the hardware
+devices and, if a device has been replaced, we need to be able to retrieve the
+new information and update the hardware inventory to provide a constant
+up-to-date vision on the status of the devices.
+
+### Goals
+
+1. Decide which information on the devices to retrieve.
+2. Agree on how to retrieve such information.
+3. Provide a way to manually/automatically collect and update the devices
+information.
+
+### Non-Goals
+
+1. Any update on device data should not trigger any reaction from ironic.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+- Depending on the device, we need to collect the following information, when
+applicable:
+    * Capacity
+    * Location
+    * Manufacturer
+    * Model
+    * Part Number
+    * Serial Number
+    * Status
+    * SMART details
+
+- NVMe drives require special tools to get SMART data (e.g. nvme-cli),
+although [smartmontools](https://www.smartmontools.org/wiki/NVMe_Support)
+supports NVMe devices since version 6.5.
+
+- [Hdparm](https://en.wikipedia.org/wiki/Hdparm) can provide SMART data for
+most of drives. It doesn't work with NVMe devices.
+
+- [python-hardware](https://github.com/redhat-cip/hardware) can collect SMART
+data, although it requires code changes to detect and provide info on NVMe
+devices, for example relying directly on nvme tools instead of smartmontools.
+
+- [Ironic Python Agent (IPA)](https://github.com/openstack/ironic-python-agent)
+is already able to collect most of this information.
+Using a subset of the IPA code running inside a container in the hosts where we
```

Comment:
Do we need to talk about what sort of privileges the container will have to run with to collect the information we need? (net=host at a minimum)

Reply:
@zaneb we can mention that without going into too much details

Reply:
I think we should add at least a little detail here to say that the host may need escalated privileges.

Reply:
I've added a separate line at L98 to add this detail